### PR TITLE
Fix panel toolbar hover to reveal.

### LINKED ISSUE #2377: All panel headers appear when hovering over any panel in a tab
**Description**

Create a top-level Tab panel. Put two Image panels inside of a single tab. Hover over one of the Image panels, and observe the Image panel header appearing on both Image panels.

### PR DESCRIPTION
**Pull request by foxymiles**
**User-Facing Changes**
This improves the hover interaction to hide & show toolbars in tab panels.

**Description**
This supplies a new `usePanelMousePresence` hook that tracks the presence of the mouse within the parent panel that can be used by child components to control their visibility.

Fixes #2377 

https://user-images.githubusercontent.com/93935560/149392145-6027c167-142f-4bb6-ab91-9a6aeea77c00.mov

### CODE REVIEW

File: packages/studio-base/src/components/PanelRoot.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import styled from "styled-components";
+
+import { colors, spacing } from "@foxglove/studio-base/util/sharedStyleConstants";
+
+// This is in a separate file to prevent circular import issues.
```
Comment (defunctzombie) [inline packages/studio-base/src/components/PanelRoot.tsx:9]: Circular import issues with what?

File: packages/studio-base/src/hooks/usePanelHoverVisibility.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  ref: MutableRefObject<HTMLElement | ReactNull>,
+  options?: { forceVisible: boolean },
+): void {
+  const listener = useCallback(
+    (e: MouseEvent) => {
+      if (!ref.current) {
+        return;
+      }
+
+      if (e.type === "mouseenter") {
```
Comment (defunctzombie) [inline packages/studio-base/src/hooks/usePanelHoverVisibility.tsx:24]:
```suggestion
      if (ev.type === "mouseenter") {
```

File: packages/studio-base/src/hooks/usePanelHoverVisibility.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { MutableRefObject, useCallback, useEffect } from "react";
+
+import { PanelRoot } from "@foxglove/studio-base/components/PanelRoot";
+
+/**
+ * Attaches a listener to the parent panel mouse enter & mouse leave events.
+ *
+ * @param ref The element to hide and show on panel hover.
```
Comment (defunctzombie) [inline packages/studio-base/src/hooks/usePanelHoverVisibility.tsx:12]: Should this return the visibility and allow the caller to show/hide their element themselves? So this hook exists to manage the enter/leave events.

File: packages/studio-base/src/hooks/usePanelHoverVisibility.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ */
+export function usePanelHoverVisibility(
```
Comment (defunctzombie) [inline packages/studio-base/src/hooks/usePanelHoverVisibility.tsx:14]:
```suggestion
export function usePanelMousePresence(
```

If you go with the approach of it returning a state vs changing the element directly

File: packages/studio-base/src/components/PanelRoot.tsx (hunk 1 above)
Comment (foxymiles) [reply packages/studio-base/src/components/PanelRoot.tsx:9]: For example the toolbar imports the hook which imports `PanelRoot` which imports the toolbar.

File: packages/studio-base/src/hooks/usePanelHoverVisibility.tsx (hunk 4 above)
Comment (foxymiles) [reply packages/studio-base/src/hooks/usePanelHoverVisibility.tsx:14]: Good idea. I'll change it.

File: packages/studio-base/src/hooks/usePanelHoverVisibility.tsx (hunk 3 above)
Comment (foxymiles) [reply packages/studio-base/src/hooks/usePanelHoverVisibility.tsx]: Yes good idea. I've reworked it to just return a boolean.

**Pull request merged**